### ISSUE TITLE
Implement closing orders (GERS finalization) (prior to V12)

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1063]
_Imported from trac ticket 1063.
Created by **kbower**
Opened in trac: Sep 14, 2011, 12:19:29 PM
Last modified in trac: Jan 15, 2013, 3:47:35 PM_
***
See also #587

For V12, see [#1962](1962)

~~Blocked by http://www.sqlalchemy.org/trac/ticket/2449~~

Additional checklist
 * check why delay at isf
 * ~~Fix problem with cloned lines of closed orders~~
 * ~~Will need to fix mockdata load~~
 * ~~If all CustomerOrder.orders are != 'O', then do not even run discount/tax/shipping logic anymore... at that point it is 'etched in stone'.~~
 * For autoupdates at save:
   * ~~add (some) meaningful comments~~
   * ~~go through all _set_defaults, _validate, _system_updates and factor which functionality must also be run for save of autoadjustment~~
 * ~~For autoadjustment lines, make restricted attributes work as intended.~~
 * ~~Verify balance is correct (before/after save)~~ (sj)
 * ~~Don't allow double click to change price if restricted attribute~~
 * ~~Don't allow add/remove/move lines if restricted attribute~~

***
* kbower attached [sof.txt](https://trac.retailarchitects.com/trac/attachment/ticket/1063/sof.txt) on Oct 23, 2012, 4:43:39 PM
* kbower attached [sqla.py.manually_add_unless_deleted](https://trac.retailarchitects.com/trac/attachment/ticket/1063/sqla.py.manually_add_unless_deleted) on Oct 30, 2012, 12:52:52 PM
* kbower attached [sqla.py.readd_expunged](https://trac.retailarchitects.com/trac/attachment/ticket/1063/sqla.py.readd_expunged) on Oct 30, 2012, 12:53:06 PM
* kbower attached [objects.py.delayclonerelationships](https://trac.retailarchitects.com/trac/attachment/ticket/1063/objects.py.delayclonerelationships) on Oct 30, 2012, 12:53:58 PM

**Comment by kbower** [comment]
_**kbower** trac comment on Sep 14, 2011, 12:19:34 PM:_

Dispatcher does close orders but we operate quite differently than Kanes and El Dorado given that a large percentage of the orders run through dispatcher are IST’s and the actual sales orders are handled at the store level so those are not closed by dispatcher. I would like to see closing an order move up the priority ladder. I think for our model it would be nice so that our office staff in the stores is not having to work in both systems as much. It’s not an urgent thing but I do think it should happen as soon as ya’ll finish backorders, installments, MDB, MCR.

Thanks

Sal

**Comment by kbower** [comment]
_**rneville** trac comment on Oct 21, 2011, 8:21:14 PM:_

Sal's telling me this is not a show-stopper right now, but will quickly become so. They do a very large percentage of take-with / accessory orders that require immediate finalization. Therefore it's not really a back office function for them..

**Comment by kbower** [comment]
_**kbower** trac comment on Mar 16, 2012, 1:53:01 PM:_

Gentlemen,
    In an effort to get the most complete answer possible, I'd ask *each* of you to please reply:

Please list everything that most occur to close an order.

Thank you,
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Apr 25, 2012, 10:39:55 AM:_

1-The status of the AR tran goes from anticipated to temporary

2-The post date of the AR tran is the date the order is closed

3-Inventory

4-You can still make comments to a finalized sale but I don't think you can change anything else to a finalized sale

If I think of more stuff I will send it your way

Sal

**Comment by kbower** [comment]
_**kbower** trac comment on Apr 25, 2012, 10:40:57 AM:_

Sorry, I thought I had answered this question previously.  I didn't:

By "closing" I am assuming you mean "Finalizing" in GERS terms.

This is what occurs:

AR_TRN table is updated for the specific DEL_DOC_NUM.

POST_DATE is updated

STAT_CD is updated to "T"

SO table is updated

                FINAL_DT  is filled

                STAT_CD is updated to "F"

ITM_TRN table transactions are created.

ITT_CD = 'SOF' for each SO_LN

Financed Transactions may be created for submission to the finance company, but  not sure how it works now with LOFT.

I am not aware of other transactions or updates.

Thanks!

Jesús R. Capó, VP, CIO

**Comment by kbower** [comment]
_**kbower** trac comment on Apr 25, 2012, 10:59:15 AM:_

Oh sorry, just that the inventory count is updated to reflect the item being in the customers possession and no longer in warehouse or showroom stock.

*From:*Kent Bower [mailto:kb@retailarchitects.com]
*Sent:* Wednesday, April 25, 2012 9:40 AM
*To:* Sal Genovese
*Subject:* Re: #1063: Closing orders

What is meant by #3 "Inventory"?

**Comment by kbower** [comment]
_**kbower** trac comment on Apr 26, 2012, 7:16:25 AM:_

Kent,

Good morning, when you finalize a sale the following should happen:

1.        Verify all inventory skus have SO_LN.STORE_CD, SO_LN.LOC_CD filled
2.        Verify that for all inventory skus SO_LN.FIFO_CST is not null
3.        Prompt for SO.FINAL_DT
4.        Prompt for SO.FINAL_STORE_CD
5.        SO.STAT_CD is updated to 'F'
6.        SO.FINAL_DT is updated to the final date entered which cannot be before the last closed day or greater than today's         date
7.        ITM_TRN SOF inserts made for each inventory item on the order
8.        AR_TRN.POST_DT is updated to SO.FINAL_DT
9.        AR_TRN.STAT_CD is updated to 'T'
10.        Remove links to SO_LN$PO_LN
11.        Remove links to IST_LN$SO_LN
12.        Insert SO.DEL_DOC_NUM, CO_CD, POST_DT, ORIGIN_CD into GL_DOC_TRN if the company is doing auto posting
13.        Kanes Furniture would like to notify the user of any outstanding balance on the sale and give option to make a payment         or dollar transfer

That's all I have for now if I remember anything else I will email you.

Thanks,

Wayne Liburd

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2012, 8:47:26 AM:_

Question posed to Kanes and El Dorado:
_What happens when you finalize an order (or attempt to) that has an unfilled line? _

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2012, 8:51:25 AM:_

Thanks, Tim.

-------- Original Message --------
Subject:         RE: closing an order with unfilled line
Date:         Tue, 23 Oct 2012 12:51:44 +0000
From:         Tim Finley <tfinle@kanesfurniture.com>
Reply-To:         Tim Finley <tfinle@kanesfurniture.com>
To:         kb@retailarchitects.com <kb@retailarchitects.com>



GERS will not allow an order to be finalized if the locations are not filled. This applies to CRM's and Exchanges as well.

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2012, 9:01:46 AM:_

I understand that GERS won't allow a ticket to be closed unless all the reservable lines are filled.

Is a reservable line determined by:
a) ITM.INVENTORY = 'Y'
b) ITM.ITM_TP_CD
c) some other criteria?

Thanks,
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2012, 9:24:38 AM:_

Doug,
    Thanks for the prompt response!  Unfortunately, Loft (also) sets this flag.  Since Loft is setting it, it doesn't help me answer the original question (circular question).

I presume the answer is that for *itm.inventory='Y'*, the line must be filled in order to finalize the ticket, but I need verification if so (and any additional information related to closing these lines).

Thanks again,
Kent

----
On 10/23/2012 9:17 AM, Doug Warner wrote:
Hi Kent,
I would look at the PICKED field in SO_LN table. Then you don't need to look at the ITM table.
Values and what they mean:
F  when store_cd, loc_cd, fill_dt, fifo_cst are not null and void_flag = 'N'  or item is *reverved
*V when void_flag ='Y', *Partial Void or line is voided*
N when store_cd, loc_cd are not null, void_flag ='N', but filled_dt is null (*only occurs on cash and carry lines*)
Z when store_cd, loc_cd, fill_dt are null and void_flag ='N' **
X similar to Z, but for a non-inventory sku (*itm.inventory='N'*)
S when item is a serviced item (*customer owned merchandise*)
Hope this helps.
Doug
**

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2012, 9:34:53 AM:_

Kent,

Good morning, if ITM.INVENTORY = 'Y' then SO_LN.STORE_CD and SO_LN.LOC_CD must not be null in order to finalize the sale.

Thanks,

Wayne Liburd

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2012, 2:05:34 PM:_

Replying to [kbower](comment:25):
>...
>Financed Transactions may be created for submission to the finance company, but  not sure how it works now with LOFT.
>...
>Thanks!
>
>Jesús R. Capó, VP, CIO
>


Scott, we could do this in a rule for El Dorado if they want to submit financing at the time of finalization, but instead of the somewhat brittle async, this would be better suited for a process daemon.

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 23, 2012, 2:52:37 PM:_

Wayne,
1)   What should ideally occur if an inventory sku has a null
SO_LN.FIFO_CST?  (This isn't something just anyone can fix, nor can they
fix it through Loft, so for now we will probably just return an error and
not close the order.)
   Also, please explain why this is needed and why sometimes it might be
null.

2)  Regarding: "Insert SO.DEL_DOC_NUM, CO_CD, POST_DT, ORIGIN_CD into
GL_DOC_TRN if the company is doing auto posting", how does Loft determine
if the company is doing auto posting?

Thanks!
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 26, 2012, 7:35:46 AM:_

Kent,

Good afternoon, a null SO_LN.FIFO_CST is a rare occurrence and probably an error is occurring when reserving a line unless someone updated the SO_LN outside of the application.

There is an Accounting system parameter '25. Enable automatic posting of Inventory and Sales?                    : N' that would be needed to set to 'Y'.

Thanks,

*Wayne Liburd*

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 26, 2012, 7:43:03 AM:_

Unless your company uses the below piece of functionality, we will not (at this time) implement it as part of closing (finalizing) orders.

Please reply if your company has this Accounting system parameter set to 'Y':
‘25. Enable automatic posting of Inventory and Sales?'

If no one replies, I will assume this functionality is not currently needed in Loft.

Thanks!
Kent

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 26, 2012, 9:28:19 AM:_

Ok. Sure, Sal.  Be difficult.  ;)

Any chance you know the specifics of what gets recorded into GL_DOC_TRN exactly when a ticket is closed?  If not, could you run a test on TRAIN (GERS) and let me know the order id you closed?

Thanks!

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 26, 2012, 10:46:00 AM:_

From:         Sal Genovese <sal@ivansmith.net>
To:         'Kent Bower' <kb@retailarchitects.com>


Sorry, but glad miskelly didn't leave me hanging!

I just finalized so# 0923131ENVO on TRAIN

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 26, 2012, 12:59:43 PM:_

```
SQL> desc GL_DOC_TRN
 Name                                       Null?    Type
 ------------------------------------------ -------- -----------------------------
 DOC_NUM                                    NOT NULL VARCHAR2(14)
 CO_CD                                      NOT NULL VARCHAR2(3)
 POST_DT                                    NOT NULL DATE
 ORIGIN_CD                                  NOT NULL VARCHAR2(5)
 PROCESSED                                  NOT NULL VARCHAR2(1)


SQL> set lines 22222
SQL> set trimspool on
set trimout on
SQL> SQL>
SQL> select * from GL_DOC_TRN where DOC_NUM ='0923131ENVO';

DOC_NUM        CO_ POST_DT              ORIGI P
-------------- --- -------------------- ----- -
0923131ENVO    01  26-OCT-2012 00:00:00 SOM   Y

```

**Comment by kbower** [comment]
_**kbower** trac comment on Oct 26, 2012, 1:05:35 PM:_

Sal,
    I would assume that 'PROCESSED' flag was 'N' when it was first inserted, but it is 'Y' now.


SQL> select * from GL_DOC_TRN where DOC_NUM ='0923131ENVO';

DOC_NUM        CO_ POST_DT              ORIGI P
-------------- --- -------------------- ----- -
0923131ENVO    01  26-OCT-2012 00:00:00 SOM   Y

Do you know whether it should insert as 'Y' or 'N'?

**Comment by kbower** [comment]
_**rneville** trac comment on Jan 11, 2013, 10:03:54 AM:_

pulling forward

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 15, 2013, 3:47:35 PM:_

(Changeset [[3553]](../commit/a9a0f0e2e48cc3a084f7ebd293de9c1c20bebd8e))
* closes #1063, Implement closing orders (GERS finalization) (prior to V12)

**Comment by kbower** [comment]
_**kbower** trac status change on Jan 15, 2013, 3:47:35 PM:_ **closed**
